crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -2625,7 +2636,18 @@ impl<'db> TypeInferenceBuilder<'db> {
                     if !dunder_class_getitem_method.is_unbound() {
                         return dunder_class_getitem_method
                             .call(self.db, &[slice_ty])
-                            .unwrap_with_diagnostic(self.db, value.as_ref().into(), self);
+                            .return_ty_result(self.db, value.as_ref().into(), self)
+                            .unwrap_or_else(|err| {
+                                self.add_diagnostic(
+                                    (&**value).into(),
+                                    "call-non-callable",
+                                    format_args!(
+                                        "Method `__class_getitem__` is not callable on object of type '{}'.",
```

Comment:
Same comment as above, let's include the type we inferred for `__class_getitem__`.